abci/client/local_client.go

```diff
@@ -15,6 +15,12 @@ import (
 type localClient struct {
 	service.BaseService
 
+	// This mutex originally protected the ABCI machinery from reentering the
+	// Application.  However, overuse of it caused performance problems for
+	// read-only queries which the Application (say, Cosmos SDK) can perfectly
+	// well handle in parallel.  So, where we indicate that the mutex isn't used,
+	// it is because the ABCI message does not result in any mutations of the
+	// KVStore (or other state), and thus can be handled concurrently.
```

Comment:
I believe that this should probably be clarified in the specification and updated in the Upgrading/Changelog. 

It's a bit hard to imagine that if we need to protect the application from concurrent write operations, that allowing read operations to be concurrent with write operations will *not* lead to corruption or incorrect data. At the same time, I think the whole mutex is maybe a bit of an overreach and maybe it should be removed entirely, and we should rely on the applications to control to concurrency.  

Perhaps it makes sense to parameterize this so that there can be "mutexed clients" and "non-mutexed clients" flavors that application developers can use? That would definitely give me more confidence in this. 

I also wonder if it makes the most sense to couple this change with a breaking change for ABCI, so that application authors (SDK and otherwise) can reckon with the implications of the change in concurrency model.

Reply:
Thanks for the review!

> Perhaps it makes sense to parameterize this so that there can be "mutexed clients" and "non-mutexed clients" flavors that application developers can use? That would definitely give me more confidence in this.

That was my conclusion, too, after doing this experiment.  I think I'll work out a new PR with that approach, and see if I can get some traction with it.

Reply:
should we close this one in the mean time?

Reply:
> should we close this one in the mean time?

I redid the current PR to provide a `committingClient` for things like the SDK which don't actually write data until certain commit points.  That seems to provide the best balance between existing ABCI behaviour and improving the query performance.

I tested this approach in the SDK PR cosmos/cosmos-sdk#10089 .

I had tried earlier with a "non-mutexed client" and then implementing fine-grained locking in the SDK's `baseapp`, but that ran into problems, not least of which was making it hard to override methods in the `baseapp` if it provided its own locking.